Source/main.cpp

```diff
@@ -434,66 +461,98 @@ static void MX_GPIO_Init(void)
-    //   input[2*i] = ADC1ConvertedValues[i];
-    //   input[2*i+1] = 0; //Imaginary part set to zero
-    // }
-    /**
-     * @warning MUST BE UPDATED!!
-     */
-    data_hyd_port[i] = 0;
-    data_hyd_starboard[i] = 0;
-    data_hyd_stern[i] = 0;
+/**
+ * @brief Read the ADC by using DMA
+ * 
+ * @param p_data_hyd_port Pointer to the memory for the port hydrophone
+ * @param p_data_hyd_starboard Pointer to the starboard hydrophone's memory
+ * @param p_data_hyd_stern Pointer to the stern hydrophone's memory
+ * 
+ * @warning Unsure if I have read/understood it correctly!
+ * Assuming that there exists two possibilities for the DMA to push the
+ * data depending on the rank:
+ * 
+ *    1) 
+ *      Data is pushed in serial after rank. Since the DMA/ADC is in scan
+ *      mode, it scans all of the channel after rank. Most likely
+ *        
+ *        r1 = rank1, r2 = rank2, r3 = rank3
+ *        {r1, r2, r3, r1, r2, r3, ... , r1, r2, r3}
+ * 
+ *    2) 
+ *      All of the data for the highest priority rank is pushed first, then
+ *      the data belonging to the second priority rank and so forth
+ * 
+ *      {r1, r1, ..., r1, r2, r2, ..., r2, r3, r3, ..., r3}
+ * 
+ * From my understanding of the circular DMA and ADC in scan-mode, it uses
+ * the first method. This is assumed true, however that is potentially a serious
+ * bug! 
+ */
+static void read_ADC(float32_t* p_data_hyd_port, float32_t* p_data_hyd_starboard,
+            float32_t* p_data_hyd_stern){
+  /**
+   * Reading the data. Dropping the last couple of datapoints, since
+   * 4096 % 3 = 1. Reducing the number of datapoints reduces the 
+   * accuracy of the analysis, however prevents out-of-range error
+   */
+  for(int i = 0; i < DSP_CONSTANTS::DMA_BUFFER_LENGHT - 
+        NUM_HYDROPHONES, i += NUM_HYDROPHONES){
+    data_hyd_port[i] = ADC1ConvertedValues[i];
+    data_hyd_starboard[i] = ADC1ConvertedValues[i + 1];
+    data_hyd_stern[i] = ADC1ConvertedValues[i + 2];
   }
 }
 
-/* USER CODE END 4 */
 
 /**
-  * @brief  This function is executed in case of error occurrence.
-  * @retval None
-  */
-void Error_Handler(void){
-  /* USER CODE BEGIN Error_Handler_Debug */
-  /* User can add his own implementation to report the HAL error return state */
-
-  /* USER CODE END Error_Handler_Debug */
-}
-
-/**
- * @brief Function to handle coordination with the through ethernet
+ * @brief Function to handle communication over ethernet
  * 
  * @warning Not implemented as of 11.12.2020, as I am not sure how the
  * communication between the STM32 and the main CPU should be
- * 
  */
 uint8_t ethernet_coordination(void){
   return 0;
 }
 
 
+/**
+ * @brief Function to log the errors
+ * Could also be interesting to implement a timestamp to detect when the
+ * error occurs. This could be done in the future, but is not a priority
+ * as of 14.12.2020
+ * 
+ * NOTE: Could be improved by using a txt-file or other log, or send the number
```

Comment:
I agree, logging the errors on the Xavier seems to be a good option. Having the time synced with the Xavier would also allow us to pinpoint where eventual errors appear, in relation to everything else that is going on in our system.

Reply:
But yeah, as you write, this is still early phase, so I think figuring out the communication protocol between the Xavier and the board is the main priority in that regard.